ionic-team/stencil
# feat(runtime): automatically insert `key` attrs during compilation

Putting up the code for this now so we can start looking at it!

## What is the current behavior?

We have a number of issues in the backlog relating to shortcomings in how children are reconciled across virtual DOM re-renders at present. For example, #1968 is caused by an interaction between a failure to correctly recognize the same child across a re-render and another issue with slot content relocation in a `scoped` context.

This implements an approach to addressing a subset of these issues by automatically inserting `key` attributes at compile-time, making it easier at run-time for the virtual DOM to determine when two children of a virtual DOM node should be matched up with each other.


## What is the new behavior?

Now as part of transpilation Stencil will insert randomly-generated `key` attributes into JSX syntax tree nodes. This allows the more-accurate key-based child reconciliation algorithm to be used almost all the time, fixing some edge cases where previously the Stencil runtime would accidentally misplace a child, failing to, for instance, correctly match up an old and new vdom node when nodes were reordered.

- fixes #1968 
- fixes #5263 

## Does this introduce a breaking change?

I don't believe it does!




## Testing

I have added some tests which exercise the change and also ensured that all of our existing tests are passing without issue.

I also built and ran this in Framework, where all of the unit tests are passing and some smoke testing of components (the datetime in particular) doesn't reveal any potential problems.

But! For testing this I think it would be a good idea to check out the reproduction case from   #1968. Then:

- confirm you can reproduce the issue with `@stencil/core@latest`
- checkout this branch
- `npm run build && npm pack`
- install the tarball in the reproduction case
- confirm that the issue is fixed
